presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java

```diff
@@ -49,34 +60,200 @@ public JoinStatsRule(FilterStatsCalculator filterStatsCalculator)
+    {
+        PlanNodeStatsEstimate innerJoinStats = computeInnerJoinStats(node, leftStats, rightStats, session, types);
+        PlanNodeStatsEstimate rightAntiJoinStats = calculateAntiJoinStats(node.getFilter(), flippedCriteria(node), rightStats, leftStats);
+        return addAntiJoinStats(innerJoinStats, rightAntiJoinStats, getRightJoinSymbols(node));
+    }
+
+    private PlanNodeStatsEstimate computeInnerJoinStats(JoinNode node, PlanNodeStatsEstimate leftStats, PlanNodeStatsEstimate rightStats, Session session, Map<Symbol, Type> types)
+    {
+        List<Expression> comparisons = node.getCriteria().stream()
                 .map(criteria -> new ComparisonExpression(EQUAL, criteria.getLeft().toSymbolReference(), criteria.getRight().toSymbolReference()))
                 .collect(toImmutableList());
+        Expression predicate = combineConjuncts(combineConjuncts(comparisons), node.getFilter().orElse(TRUE_LITERAL));
+        PlanNodeStatsEstimate crossJoinStats = crossJoinStats(node, leftStats, rightStats);
+        return filterStatsCalculator.filterStats(crossJoinStats, predicate, session, types);
+    }
+
+    @VisibleForTesting
+    PlanNodeStatsEstimate calculateAntiJoinStats(Optional<Expression> filter, List<JoinNode.EquiJoinClause> criteria, PlanNodeStatsEstimate leftStats, PlanNodeStatsEstimate rightStats)
+    {
+        if (filter.isPresent()) {
+            // non-equi filters are not supported
+            return UNKNOWN_STATS;
+        }
+
+        PlanNodeStatsEstimate outputStats = leftStats;
 
-        PlanNodeStatsEstimate mergedInputCosts = crossJoinStats(joinNode, lookup, session, types);
-        Expression predicate = combineConjuncts(combineConjuncts(comparisons), joinNode.getFilter().orElse(TRUE_LITERAL));
-        PlanNodeStatsEstimate filteredStatistics = filterStatsCalculator.filterStats(mergedInputCosts, predicate, session, types);
+        for (EquiJoinClause clause : criteria) {
+            SymbolStatsEstimate leftColumnStats = leftStats.getSymbolStatistics(clause.getLeft());
+            SymbolStatsEstimate rightColumnStats = rightStats.getSymbolStatistics(clause.getRight());
 
-        PlanNodeStatsEstimate.Builder finalStatsBuilder = PlanNodeStatsEstimate.builder()
-                .setOutputRowCount(filteredStatistics.getOutputRowCount());
-        filteredStatistics.getSymbolsWithKnownStatistics().stream()
-                .filter(node.getOutputSymbols()::contains)
-                .forEach(symbol -> finalStatsBuilder.addSymbolStatistics(symbol, filteredStatistics.getSymbolStatistics(symbol)));
+            StatisticRange rightRange = new StatisticRange(rightColumnStats.getLowValue(), rightColumnStats.getHighValue(), rightColumnStats.getDistinctValuesCount());
+            StatisticRange leftRange = new StatisticRange(leftColumnStats.getLowValue(), leftColumnStats.getHighValue(), leftColumnStats.getDistinctValuesCount()).subtract(rightRange);
```

Comment:
`leftRange` variable name is misleading. It is `antiRange` or `leftMinusRightRange`.